aws/resource_aws_iam_instance_profile.go

```diff
@@ -160,7 +160,16 @@ func instanceProfileAddRole(iamconn *iam.IAM, profileName, roleName string) erro
 		RoleName:            aws.String(roleName),
 	}
 
-	_, err := iamconn.AddRoleToInstanceProfile(request)
+	err := resource.Retry(30*time.Second, func() *resource.RetryError {
+		var err error
+		_, err = iamconn.AddRoleToInstanceProfile(request)
+		// Todo: handle retryable and non-retryable errors
```

Comment:
Hi @joakiboxxx! Thanks for submitting this! Can you please update this to the below (including the comment):

```go
// IAM unfortunately does not provide a better error code or message for eventual consistency
// InvalidParameterValue: Value (XXX) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name
if isAWSErr(err, "InvalidParameterValue", "Invalid IAM Instance Profile name") {
  return resource.RetryableError(err)
}
if err != nil {
  return resource.NonRetryableError(err)
}
return nil
```

Reply:
Hi Brian!

I applied your patch and tested it, and the change no longer solves my problem ;) With a debug inside the Retry callback, I get this from err:

[DEBUG] Error adding role NoSuchEntity: The role with name joakiboxxx_role cannot be found.

Wouldn't you agree that NoSuchEntity is indeed a retryable error? If so, may I propose the following:

if isAWSErr(err, "InvalidParameterValue", "Invalid IAM Instance Profile name") || isAWSErr(err, "NoSuchEntity", "The role with name") {
  return resource.RetryableError(err)
}